src/Microsoft.ML.Data/Data/DataViewUtils.cs

```diff
@@ -942,8 +942,10 @@ public override void Set(BatchColumn batchCol)
                     public override void Unset()
                     {
                         Contracts.Assert(_index <= _count);
-                        if (Values != null)
-                            _pool.Return(Values);
+                        // Remove all the objects from the pool
+                        // to free up references to those objects
+                        while (_pool.Count > 0)
+                            _pool.Get();
```

Comment:
I believe the original code was a copy paste error. If you look at the code for `Set`, it has the same if check and the same `_pool.Return..`. `_pool.Return` actually adds it to the Bag (bad method name? Not intuitive). So I think your fix maxes sense since get is actually removing it from the pool unlike return.

Right now since you are just doing a get, the resources are freed when the GC is run. Would it be better to do something like this?
```
while (_pool.Count > 0)
{
    var item = _pool.Get();
    (item as IDisposable)?.Dispose();
}
```
This way if it is disposable it will free it up right away. Not sure if that would impact performance much though.

What are your thoughts?

Reply:
We don't have many disposable objects. The disposable objects we consume are mostly from Tensorflow and OnnxRuntime. These aren't part of the object pool. If we add that line, I can't think of a test we can add that would exercise that code path.

---
In reply to: [534439391](https://github.com/dotnet/machinelearning/pull/5521#discussion_r534439391) [](ancestors = 534439391)